[wip] Adding additional fields

brianloveswords opened the pull request:
for issue #101 
# All the fields

Issuers have Programs, Programs have Badges.
Badges can count towards earning other Badges. Some badges can only be earned by earning prerequisite Badges. Badges can count towards the earning of another Badge in specific numbers, so a single Badge can count as 2 points toward another Badge, or 1 point, or zero...etc.
## Organization / Issuer
- [x] Issuer Name
- [x] Issuer Description - _mockups don't show this anywhere_
- [x] Issuer Logo - _mockups don't show this anywhere_
- [x] Issuer URL
## Program
- [x] Program Name
- [x] Program Description
- [x] Begin Date
- [x] End Date
- [x] Program site, a url where more information lives.
- [x] Program Logo
- [x] Phone Number - _mockups don't show this anywhere_
- [x] Contact - _mockups don't show this anywhere_
## Badge
- [x] Badge Name
- [x] Badge Description
- [x] Badge Criteria
- [x] Badge Category
- [x] Tags
- [x] Time to earn (how long will it take to earn this badge? 1-4 hours, 1 day, 1 week)
- [x] Age Range (<13, 13-18, 19-24)
- [ ] Does this badge have prerequisites?
- [ ] Does this badge count towards another badge? Which badge? How much does it count against that badge?
- [x] Badge type (enumerated list : self-paced, in person, etc)

TODO
- [ ] Add the fields to the models
- [ ] Add the fields to the admin panel
- [ ] Add the fields to API

brianloveswords committed:
Additional data fields programs

brianloveswords committed:
Frontend & models for (some) extra badge fields

This is still missing the prereqs stuff, but this adds age range, time
to earn, category and badge type to the backend models and the
frontend. This also had some refactoring of the badge `create` and
`update` endpoints to pull out common code.

stenington commented on template.js:
```diff
@@ -61,5 +61,15 @@ exports.buildEnvironment = function buildEnvironment(options) {
   });
 
+  env.addFilter('dateInput', function (date) {
+    return date.toISOString().split('T')[0];
```
This needs to deal with null dates, if start date and end data aren't required when defining a program. Otherwise I get 

```
Express
500 Template render error: (/Users/mlarsson/Work/checkouts/openbadger/views/admin/create-or-edit-issuer.html) Template render error: TypeError: Object has no method 'toISOString'
```

trying to edit a program for which I didn't choose dates.

brianloveswords replied:
Start and end will be required (I think?), but I'll still add a null check just to be extra defensive.

brianloveswords replied:
IF ONLY JAVASCRIPT HAD `Maybe`, [AMIRITE](http://nickknowlson.com/blog/2013/04/16/why-maybe-is-better-than-null/)

cmcavoy replied:
Not all programs have a start and end date.

On Wed, May 22, 2013 at 1:05 PM, Brian J Brennan
notifications@github.comwrote:

> In template.js:
> 
> > @@ -61,5 +61,15 @@ exports.buildEnvironment = function buildEnvironment(options) {
> >      }[access];
> >    });
> > -  env.addFilter('dateInput', function (date) {
> > -    return date.toISOString().split('T')[0];
> 
> Start and end will be required (I think?), but I'll still add a null check
> just to be extra defensive.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mozilla/openbadger/pull/110/files#r4344895
> .

## 

@chmcavoy http://twitter.com/chmcavoy
http://lonelylion.com

brianloveswords committed:
Protect against nulls.

brianloveswords committed:
Add images to issuers.

This also does a lot of work to simplify things on the backend. Instead
of processing programs at the same time we try to process issuers, it is
a separate step. It's (arguably) a worse user experience on the
frontend, but A) the UX is in flux right now so we don't know what it's
going to look like anyway, and B) it's *way* simpler on the backend.

brianloveswords committed:
Can add and edit programs again.

This time with more images and other fields and stuff!

brianloveswords committed:
Add programs, issuers to the api
